module/x/gravity/keeper/batch.go

```diff
@@ -1,300 +1,362 @@
 package keeper
 
 import (
-	"fmt"
+	"crypto/sha256"
 	"strconv"
 
+	tmbytes "github.com/tendermint/tendermint/libs/bytes"
+
 	"github.com/cosmos/cosmos-sdk/store/prefix"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/ethereum/go-ethereum/common"
 
 	"github.com/cosmos/gravity-bridge/module/x/gravity/types"
 )
 
-const OutgoingTxBatchSize = 100
-
-// BuildOutgoingTXBatch starts the following process chain:
+// CreateBatchTx starts the following process chain:
 // - find bridged denominator for given voucher type
 // - determine if a an unexecuted batch is already waiting for this token type, if so confirm the new batch would
 //   have a higher total fees. If not exit withtout creating a batch
 // - select available transactions from the outgoing transaction pool sorted by fee desc
 // - persist an outgoing batch object with an incrementing ID = nonce
 // - emit an event
-func (k Keeper) BuildOutgoingTXBatch(
-	ctx sdk.Context,
-	contractAddress string,
-	maxElements int) (*types.OutgoingTxBatch, error) {
-	if maxElements == 0 {
-		return nil, sdkerrors.Wrap(types.ErrInvalid, "max elements value")
+func (k Keeper) CreateBatchTx(ctx sdk.Context, contractAddress common.Address) (tmbytes.HexBytes, error) {
+	// select transfer txs from outgoing transfer tx pool sorted by fee in desc order
+	// TODO: mark the transfer txs as batched
+	txs := k.SelectTransferTxs(ctx, contractAddress)
+	if len(txs) == 0 {
+		return nil, sdkerrors.Wrapf(
+			types.ErrTxNotFound,
+			"transfer transaction pool is currently empty for token address %s", contractAddress,
+		)
 	}
 
-	lastBatch := k.GetLastOutgoingBatchByTokenType(ctx, contractAddress)
+	timeoutHeight, err := k.GetBatchTimeoutHeight(ctx)
+	if err != nil {
+		return nil, err
+	}
 
-	// lastBatch may be nil if there are no existing batches, we only need
```

Comment:
Why remove the logic which ensures that only more profitable batches are created? Has it been moved somewhere else in the code? I've attached documentation of the feature that this PR appears to remove:

> Only create more profitable batches

> This is an important design criteria for permission-less batch creation.

> Suppose we have an infinite stream of low fee spam transactions going
into the Gravity tx pool. Any given batch creation event will then lock
up high fee 'good transactions' with dozens of spam transactions.

> In this case no relayer will ever chose to relay a batch because all of
them will be unprofitable.

> In order to ensure that profitable batches are eventually created we
must avoid locking up the high fee 'good transactions' into obviously
bad batches. To add to the difficulty we don't actually know what any
token in this process is worth or what ETH gas costs.

> The solution this patch provides is to ensure that a new batch always
has higher fees than the last batch. This means that even if there are
infinite spam transactions, and infinite spamming of the permission less
create batch request batch creation will halt long enough for enough
good transactions to build up in the pool and a new more profitable
batch to be created.

> Over a long enough timescale the old batches either time out. Returning
profitable transactions to the pool to create a new even better batch. Or
the profitability of the latest batch continues to increase until one is
relayed, also freeing the previous good transactions to go into the next
batch.

> So given this condition we can say that no matter the inputs a
successful batch will eventually be created.

Reply:
I searched for all the places this function is being called in the code, and it doesn't seem like the "only more profitable" logic was moved, it was simply removed. Why make key logic changes with no ADR? How will this affect the performance and spam resistance attributes of the bridge discussed above?

Reply:
This logic is also discussed here: https://blog.althea.net/how-gravity-works

> As mentioned above, the Gravity module will only assemble a new batch if there are no batches already in the batch pool with a higher sum of fees. This is to prevent the following scenario: If there were many transactions in the transaction pool with an unprofitably low fee, and a few coming in every block with a high fee, each high fee transaction might end up in a batch with a bunch of unprofitable transactions. These batches would not be profitable to submit, and so the profitable transactions would end up in unprofitable batches, and not be submitted.

> By making it so that every new batch must be more profitable than any other batch that is waiting to be submitted, it gives the few profitable transactions that come in every block the chance to build up and form a batch profitable enough to submit.